### ISSUE TITLE
Question | Would it be a good idea to share an embedding layer between inputs and labels?

### ISSUE THREAD

**Issue by italodamato** [tensorflow/recommenders#169]
Let's say in the movielens dataset I have another column where each row is the list of the last 100 movies clicked by that user. 

<table>
<tr>
	<td>user_id
	<td>movie_watched_id
	<td>last_100_movies_clicked_ids
<tr>
	<td>'123'
	<td>'67865'
	<td>'4543534 345435435 657657'
</table>

Would it make sense to encode the `movie_watched_id` in the candidate tower using the embedding layer used in the query tower to embed `last_100_movies_clicked_ids`?

Like, having this user model:
```
class UserModel(tf.keras.Model):

  def __init__(self):
    super().__init__()

    self.user_embedding = tf.keras.Sequential([
        tf.keras.layers.experimental.preprocessing.StringLookup(
            vocabulary=unique_user_ids, mask_token=None),
        tf.keras.layers.Embedding(len(unique_user_ids) + 1, 32),
    ])

    self.last_100_movies_clicked_ids_embedding = tf.keras.Sequential([
        tf.keras.layers.experimental.preprocessing.TextVectorizer(
            vocabulary=unique_last_100_movies_clicked_ids, mask_token=None),
        tf.keras.layers.Embedding(len(unique_last_100_movies_clicked_ids) + 1, 32),
    ])

  def call(self, inputs):
    # Take the input dictionary, pass it through each input layer,
    # and concatenate the result.
    return tf.concat([self.user_embedding(inputs["user_id"]), 
   					  self.last_100_movies_clicked_ids_embedding(inputs["last_100_movies_clicked_ids"])
        			])
```

We define the movie model as:
```
class MovieModel(tf.keras.Model):

  def __init__(self, user_model):
    super().__init__()

    max_tokens = 10_000

    self.movie_watched_id_embedding = user_model.last_100_movies_clicked_ids_embedding

  def call(self):
    return self.movie_watched_id_embedding(inputs['movie_watched_id'])
```

**Comment by maciejkula** [comment]
Yup - this could make a lot of sense, depending on how dense your data is.

The benefits of doing something like this would be greater if you don't have a lot of data, or you have a lot of tail items. If you do have a lot of data, it might be better not to do it, as it would give your model more flexibility to learn different input and output embeddings.

**Comment by italodamato** [comment]
Interesting! What would you consider a good top-100 accuracy for a retrieval model? When can I say it's good enough to use as input for a ranking model?

**Comment by maciejkula** [comment]
That's really variable depending on your problem. If you have an existing baseline it should give you an idea of what a good offline metric is; if you don't, I'd recommend eyeballing the results together with your team and trying it out in an A/B test.

**Comment by italodamato** [comment]
I'm trying to implement the above shared embedding layer but, for some reason, my loss goes to nan on GPU (it works on CPU). I'm not sure what's causing it but it looks like a common tf issue online.

Interestingly, the following snippets should return the same result and they do... before fitting the model. Once I fit the model, the above code still returns a valid tensor while the second one is just nans. (Unfortunately, I've just run out of GPU quota and didn't get a screenshot of the nans.)

Any advice?

![image](https://user-images.githubusercontent.com/45588475/99859242-6a5d2280-2b87-11eb-914d-7f229aa89332.png)

**Comment by maciejkula** [comment]
GPUs are very different hardware with very different floating point behaviour.

I'd start with reducing the learning rate and playing with different optimizers.
